Sources/Confidence/ConfidenceValue.swift

```diff
@@ -0,0 +1,231 @@
+            return dateComponents
+        }
+
+        return nil
+    }
+
+    public func asDate() -> Date? {
+        if case let .timestamp(date) = value {
+            return date
+        }
+
+        return nil
+    }
+
+    public func asList() -> [ConfidenceValue]? {
+        if case let .list(values) = value {
+            return values.map { i in ConfidenceValue(valueInternal: i) }
+        }
+
+        return nil
+    }
+
+    public func asStructure() -> [String: ConfidenceValue]? {
+        if case let .structure(values) = value {
+            return values.mapValues { ConfidenceValue(valueInternal: $0) }
+        }
+
+        return nil
+    }
+
+    public func isNull() -> Bool {
+        if case .null = value {
+            return true
+        }
+
+        return false
+    }
+
+    public static func == (lhs: ConfidenceValue, rhs: ConfidenceValue) -> Bool {
+        lhs.value == rhs.value
+    }
+}
+
+extension ConfidenceValue {
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(value)
+    }
+}
+
+
+/// Serializable data structure meant for event sending via Confidence
+private enum ConfidenceValueInternal: Equatable, Encodable {
+    case boolean(Bool)
+    case string(String)
+    case integer(Int64)
+    case double(Double)
+    case date(DateComponents)
+    case timestamp(Date)
+    case list([ConfidenceValueInternal])
+    case structure([String: ConfidenceValueInternal])
+    case null
+}
+
+extension ConfidenceValueInternal: CustomStringConvertible {
+    public var description: String {
+        switch self {
+        case .boolean(let value):
+            return "\(value)"
+        case .string(let value):
+            return value
+        case .integer(let value):
+            return "\(value)"
+        case .double(let value):
+            return "\(value)"
+        case .date(let value):
+            return "\(value)"
+        case .timestamp(let value):
+            return "\(value)"
+        case .list(value: let values):
+            return "\(values.map { value in value.description })"
+        case .structure(value: let values):
+            return "\(values.mapValues { value in value.description })"
+        case .null:
+            return "null"
+        }
+    }
+}
+
+extension ConfidenceValueInternal {
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+
+        switch self {
+        case .null:
+            try container.encodeNil()
+        case .integer(let integer):
+            try container.encode(integer)
+        case .double(let double):
+            try container.encode(double)
+        case .string(let string):
+            try container.encode(string)
+        case .boolean(let boolean):
+            try container.encode(boolean)
+        case .date(let dateComponents):
+            let dateFormatter = DateFormatter()
```

Comment:
I would expect you to be able to use `ISO8601DateFormatter` here as well?
could work with this : `formatter.formatOptions = [.withFullDate]`

Reply:
Do you mean to serialize this to e.g. "2024-04-03T00:00:00Z"?
I think we want to keep sending "2024-04-03" for Date Values if possible (that's what the backend accepts today)

Reply:
Sorry for the confusion. No, I think you can get `YYYY-MM-DD` using `ISO8601DateFormatter` given the `dateFormatter.formatOptions = [.withFullDate]` option.